File: Code/3D Localization/code/export_deduped_suite2p.py
# -*- coding: utf-8 -*-
"""Export deduped m2 ROIs as suite2p-readable plane0 folders.

Does not modify original suite2p folders. Writes under CA1R1f5Output:

  m2_suite2p_after_dedupe/
    Plane{14,19,24,29}/suite2p/plane0/
      F.npy, Fneu.npy, spks.npy, stat.npy, iscell.npy, ops.npy
    roi_index_map.csv
    README.txt
"""
from __future__ import annotations

import csv
from pathlib import Path

import numpy as np

from dataset_config import OUT_DIR, rel_to_project
from match_linear_runner_to_recon_planes import PLANE_SPECS, PLANE_Z_BASE

KEPT_CSV = OUT_DIR / "match" / "m2_3d_coverage" / "dedupe" / "kept_rois.csv"
OUT_ROOT = OUT_DIR / "m2_suite2p_after_dedupe"


def load_kept() -> dict[int, list[tuple[int, float]]]:
    """plane_id -> [(old_roi_idx, snr), ...] sorted by old_roi_idx."""
    by: dict[int, list[tuple[int, float]]] = {pid: [] for pid in PLANE_Z_BASE}
    with open(KEPT_CSV, encoding="utf-8") as f:
        for row in csv.DictReader(f):
            pid = int(row["plane_id"])
            by.setdefault(pid, []).append((int(row["roi_idx"]), float(row["snr"])))
    for pid in by:
        by[pid].sort(key=lambda t: t[0])
    return by


def export_plane(
    plane_dir: Path, plane_id: int, kept: list[tuple[int, float]], out_plane0: Path
) -> list[dict]:
    src = plane_dir
    F = np.load(str(src / "F.npy"))
    Fneu = np.load(str(src / "Fneu.npy"))
    spks = np.load(str(src / "spks.npy"))
    stat = list(np.load(str(src / "stat.npy"), allow_pickle=True))
    iscell_src = np.load(str(src / "iscell.npy"))
    ops = np.load(str(src / "ops.npy"), allow_pickle=True).item()

    idxs = [i for i, _ in kept]
    if not idxs:
        print("plane %d: no kept ROIs, skip" % plane_id)
        return []
    for i in idxs:
        if i < 0 or i >= len(stat):
            raise IndexError("plane %d roi %d out of range" % (plane_id, i))

    F2 = np.asarray(F[idxs], dtype=np.float32)
    Fneu2 = np.asarray(Fneu[idxs], dtype=np.float32)
    spks2 = np.asarray(spks[idxs], dtype=np.float32)
    stat2 = [dict(stat[i]) for i in idxs]
    if iscell_src.ndim == 2:
        iscell2 = np.ones((len(idxs), 2), dtype=np.float64)
        iscell2[:, 1] = 1.0
    else:
        iscell2 = np.ones(len(idxs), dtype=np.float64)

    ops2 = dict(ops)
    ops2["nrois"] = int(len(idxs))

    out_plane0.mkdir(parents=True, exist_ok=True)
    np.save(str(out_plane0 / "F.npy"), F2)
    np.save(str(out_plane0 / "Fneu.npy"), Fneu2)
    np.save(str(out_plane0 / "spks.npy"), spks2)
    np.save(str(out_plane0 / "stat.npy"), np.array(stat2, dtype=object))
    np.save(str(out_plane0 / "iscell.npy"), iscell2)
    np.save(str(out_plane0 / "ops.npy"), ops2)

    rows = []
    for new_i, (old_i, snr) in enumerate(kept):
        rows.append(
            {
                "plane_id": plane_id,
                "new_roi_idx": new_i,
                "old_roi_idx": old_i,
                "snr": snr,
                "out_dir": rel_to_project(out_plane0),
            }
        )
    print(
        "plane %d: kept %d / orig %d -> %s"
        % (plane_id, len(idxs), len(stat), out_plane0)
    )
    return rows


def main() -> None:
    if not KEPT_CSV.is_file():
        raise FileNotFoundError(KEPT_CSV)
    kept_by = load_kept()
    OUT_ROOT.mkdir(parents=True, exist_ok=True)

    map_rows: list[dict] = []
    for plane_dir, plane_id, _z in PLANE_SPECS:
        plane_name = plane_dir.parents[1].name  # Plane14 etc.
        out_plane0 = OUT_ROOT / plane_name / "suite2p" / "plane0"
        map_rows.extend(
            export_plane(plane_dir, int(plane_id), kept_by.get(int(plane_id), []), out_plane0)
        )

    map_csv = OUT_ROOT / "roi_index_map.csv"
    with open(map_csv, "w", newline="", encoding="utf-8") as f:
        w = csv.DictWriter(
            f, fieldnames=["plane_id", "new_roi_idx", "old_roi_idx", "snr", "out_dir"]
        )
        w.writeheader()
        w.writerows(map_rows)

    layout_lines = ["  %s/suite2p/plane0/" % pdir.parents[1].name for pdir, _pid, _z in PLANE_SPECS]
    readme = OUT_ROOT / "README.txt"
    readme.write_text(
        "\n".join(
            [
                "Deduped method2 suite2p exports (cross-plane duplicates removed; higher SNR kept).",
                "Original suite2p folders were NOT modified.",
                "",
                "Layout:",
                *layout_lines,
                "",
                "Each plane0 contains: F.npy, Fneu.npy, spks.npy, stat.npy, iscell.npy, ops.npy",
                "(no data.bin — use the original plane's data.bin / meanImg in ops if needed).",
                "",
                "ROI indices are reindexed 0..N-1 per plane; see roi_index_map.csv for old_roi_idx.",
                "Source kept list: %s" % rel_to_project(KEPT_CSV),
                "Total kept ROIs: %d" % len(map_rows),
                "",
            ]
        ),
        encoding="utf-8",
    )
    print("wrote", map_csv)
    print("OUT_ROOT =", OUT_ROOT)
    print("total kept", len(map_rows))


if __name__ == "__main__":
    main()

File: Code/3D Localization/matlab/export_deduped_suite2p.m
function export_deduped_suite2p(cfg)
%EXPORT_DEDUPED_SUITE2P Write kept plane traces/stat sidecars under MATLAB output.
if nargin < 1, cfg = dataset_config(); end
keptCsv = fullfile(cfg.outDir, 'match', 'm2_3d_coverage', 'dedupe', 'kept_rois.csv');
outRoot = fullfile(cfg.outDir, 'm2_suite2p_after_dedupe');
if ~exist(keptCsv, 'file')
    fprintf('missing %s; skip export\n', keptCsv);
    return
end
keptBy = load_kept(keptCsv, cfg);
ensure_dir(outRoot);
mapRows = struct('plane_id', {}, 'new_roi_idx', {}, 'old_roi_idx', {}, 'snr', {}, 'out_dir', {});
k = 0;
for p = 1:numel(cfg.planeIds)
    pid = cfg.planeIds(p);
    planeDir = plane_suite2p_dir(pid, cfg);
    planeName = sprintf('Plane%d', pid);
    outPlane0 = fullfile(outRoot, planeName, 'suite2p', 'plane0');
    kept = [];
    if isKey(keptBy, pid)
        kept = keptBy(pid);
    end
    rows = export_plane(planeDir, pid, kept, outPlane0, cfg);
    for i = 1:numel(rows)
        k = k + 1;
        mapRows(k) = rows(i);
    end
end
write_csv(fullfile(outRoot, 'roi_index_map.csv'), mapRows, ...
    {'plane_id', 'new_roi_idx', 'old_roi_idx', 'snr', 'out_dir'});
readme = {
    'Deduped method2 suite2p exports (cross-plane duplicates removed; higher SNR kept).'
    'Original suite2p folders were NOT modified.'
    ''
    'Layout:'
    };
for p = 1:numel(cfg.planeIds)
    readme{end+1} = sprintf('  Plane%d/suite2p/plane0/', cfg.planeIds(p)); %#ok<AGROW>
end
readme{end+1} = ''; %#ok<AGROW>
readme{end+1} = 'Each plane0 contains numeric F.npy, Fneu.npy, iscell.npy and s2p_matlab/ sidecars.'; %#ok<AGROW>
readme{end+1} = 'ROI indices are reindexed 0..N-1 per plane; see roi_index_map.csv for old_roi_idx.'; %#ok<AGROW>
readme{end+1} = sprintf('Total kept ROIs: %d', numel(mapRows)); %#ok<AGROW>
fid = fopen(fullfile(outRoot, 'README.txt'), 'w');
fprintf(fid, '%s\n', readme{:});
fclose(fid);
fprintf('wrote %s\n', fullfile(outRoot, 'roi_index_map.csv'));
fprintf('OUT_ROOT = %s  total kept %d\n', outRoot, numel(mapRows));
end

function keptBy = load_kept(keptCsv, cfg)
keptBy = containers.Map('KeyType', 'double', 'ValueType', 'any');
for i = 1:numel(cfg.planeIds)
    keptBy(cfg.planeIds(i)) = zeros(0, 2);
end
rows = read_csv(keptCsv);
for i = 1:numel(rows)
    pid = rows(i).plane_id;
    pair = [rows(i).roi_idx, rows(i).snr];
    if isKey(keptBy, pid)
        keptBy(pid) = [keptBy(pid); pair];
    else
        keptBy(pid) = pair;
    end
end
ks = keptBy.keys;
for i = 1:numel(ks)
    arr = keptBy(ks{i});
    if ~isempty(arr)
        [~, ord] = sort(arr(:, 1));
        keptBy(ks{i}) = arr(ord, :);
    end
end
end

function rows = export_plane(planeDir, planeId, kept, outPlane0, cfg)
rows = struct('plane_id', {}, 'new_roi_idx', {}, 'old_roi_idx', {}, 'snr', {}, 'out_dir', {});
if isempty(kept)
    fprintf('plane %d: no kept ROIs, skip\n', planeId);
    return
end
s2p = load_suite2p(planeDir);
oldIdx0 = kept(:, 1); % 0-based
oldIdx = oldIdx0 + 1;
F2 = s2p.F(oldIdx, :);
Fneu2 = s2p.Fneu(oldIdx, :);
iscell2 = ones(numel(oldIdx), 2);
ensure_dir(outPlane0);
writeNPY(single(F2), fullfile(outPlane0, 'F.npy'));
writeNPY(single(Fneu2), fullfile(outPlane0, 'Fneu.npy'));
writeNPY(iscell2, fullfile(outPlane0, 'iscell.npy'));
metaSrc = fullfile(planeDir, 's2p_matlab');
metaDst = fullfile(outPlane0, 's2p_matlab');
ensure_dir(metaDst);
xpix = []; ypix = []; lam = [];
off = []; cnt = [];
med = zeros(numel(oldIdx), 2);
srcOff = readNPY(fullfile(metaSrc, 'pix_offset.npy'));
srcCnt = readNPY(fullfile(metaSrc, 'pix_count.npy'));
srcX = readNPY(fullfile(metaSrc, 'xpix.npy'));
srcY = readNPY(fullfile(metaSrc, 'ypix.npy'));
srcL = readNPY(fullfile(metaSrc, 'lam.npy'));
srcMed = readNPY(fullfile(metaSrc, 'med.npy'));
for i = 1:numel(oldIdx)
    oi = oldIdx(i);
    a = srcOff(oi) + 1;
    b = srcOff(oi) + srcCnt(oi);
    off(i, 1) = numel(xpix); %#ok<AGROW>
    cnt(i, 1) = srcCnt(oi); %#ok<AGROW>
    if srcCnt(oi) >= 1
        xpix = [xpix; srcX(a:b)]; %#ok<AGROW>
        ypix = [ypix; srcY(a:b)]; %#ok<AGROW>
        lam = [lam; srcL(a:b)]; %#ok<AGROW>
    end
    if size(srcMed, 1) >= oi
        med(i, :) = srcMed(oi, :);
    end
end
writeNPY(xpix, fullfile(metaDst, 'xpix.npy'));
writeNPY(ypix, fullfile(metaDst, 'ypix.npy'));
writeNPY(lam, fullfile(metaDst, 'lam.npy'));
writeNPY(int32(off), fullfile(metaDst, 'pix_offset.npy'));
writeNPY(int32(cnt), fullfile(metaDst, 'pix_count.npy'));
writeNPY(med, fullfile(metaDst, 'med.npy'));
copyfile(fullfile(metaSrc, 'meanImg.npy'), fullfile(metaDst, 'meanImg.npy'));
opsJ = jsondecode(fileread(fullfile(metaSrc, 'ops.json')));
opsJ.nroi = numel(oldIdx);
write_json(fullfile(metaDst, 'ops.json'), opsJ);
rel = ['m2_suite2p_after_dedupe/Plane_' num2str(planeId) '_motionfree/suite2p/plane0'];
for i = 1:numel(oldIdx)
    rows(i).plane_id = planeId;
    rows(i).new_roi_idx = i - 1;
    rows(i).old_roi_idx = oldIdx0(i);
    rows(i).snr = kept(i, 2);
    rows(i).out_dir = rel;
end
fprintf('plane %d: kept %d / orig %d -> %s\n', planeId, numel(oldIdx), numel(s2p.stat), outPlane0);
end
